From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Mio <mio@mio19.uk>
Date: Tue, 20 Oct 2026 19:24:48 +1100
Subject: [PATCH] add shell, sync daemon, and virtual keyboard
 (overlay/asteroidos)

Package the core AsteroidOS userland services on top of the middleware
layer:

- asteroid-launcher: Wayland compositor and touch smartwatch shell
- asteroid-btsyncd: Bluetooth LE synchronization daemon
- virtualkeyboard: on-screen touch keyboard with watch styling

Add the Nemo.Configuration QML plugin and the upstream libhybris
hwcomposer QPA plugin the launcher needs, and pin the launcher to the
current nightly.
---
 .../asteroidos/asteroid-btsyncd/default.nix   |  68 ++++++
 .../org.asteroidos.btsyncd.conf               |  13 ++
 .../asteroidos/asteroid-launcher/default.nix  |  82 ++++++++
 overlay/asteroidos/default.nix                |  32 +++
 .../nemo-qml-plugin-configuration/default.nix |  69 ++++++
 .../qt6-qpa-hwcomposer-plugin/default.nix     | 198 ++++++++++++++++++
 .../asteroidos/virtualkeyboard/default.nix    |  52 +++++
 7 files changed, 514 insertions(+)
 create mode 100644 overlay/asteroidos/asteroid-btsyncd/default.nix
 create mode 100644 overlay/asteroidos/asteroid-btsyncd/org.asteroidos.btsyncd.conf
 create mode 100644 overlay/asteroidos/asteroid-launcher/default.nix
 create mode 100644 overlay/asteroidos/nemo-qml-plugin-configuration/default.nix
 create mode 100644 overlay/asteroidos/qt6-qpa-hwcomposer-plugin/default.nix
 create mode 100644 overlay/asteroidos/virtualkeyboard/default.nix

diff --git a/overlay/asteroidos/asteroid-btsyncd/default.nix b/overlay/asteroidos/asteroid-btsyncd/default.nix
new file mode 100644
index 00000000..af76edf4
--- /dev/null
+++ b/overlay/asteroidos/asteroid-btsyncd/default.nix
@@ -0,0 +1,68 @@
+{
+  lib,
+  stdenv,
+  fetchFromGitHub,
+  cmake,
+  pkg-config,
+  systemd,
+  qt6,
+  qml-asteroid,
+  amber-mpris,
+  mlite,
+  kdePackages,
+  buildPackages,
+}:
+
+let
+  isCross = !stdenv.buildPlatform.canExecute stdenv.hostPlatform;
+in
+stdenv.mkDerivation (finalAttrs: {
+  pname = "asteroid-btsyncd";
+  version = "nightly";
+
+  src = fetchFromGitHub {
+    owner = "AsteroidOS";
+    repo = "asteroid-btsyncd";
+    rev = "e14fbcd9a157f0988204e874b03f6f800f89d645";
+    sha256 = "15vpd3y817gz8j1d7f56k5g1y3747grsjqc5z7b896n7rs2p8ijc";
+  };
+
+  strictDeps = true;
+
+  nativeBuildInputs = [
+    cmake
+    kdePackages.extra-cmake-modules
+    pkg-config
+    qt6.wrapQtAppsHook
+  ];
+
+  cmakeFlags = [
+    "-DECM_DIR=${kdePackages.extra-cmake-modules}/share/ECM/cmake"
+  ]
+  ++ lib.optionals isCross [
+    "-DQt6QmlTools_DIR=${buildPackages.qt6.qtdeclarative}/lib/cmake/Qt6QmlTools"
+    "-DQt6QuickTools_DIR=${buildPackages.qt6.qtdeclarative}/lib/cmake/Qt6QuickTools"
+    "-DQt6ShaderToolsTools_DIR=${buildPackages.qt6.qtshadertools}/lib/cmake/Qt6ShaderToolsTools"
+    "-DQt6LinguistTools_DIR=${buildPackages.qt6.qttools}/lib/cmake/Qt6LinguistTools"
+  ];
+
+  buildInputs = [
+    systemd
+    qt6.qtbase
+    qt6.qtdeclarative
+    qml-asteroid
+    amber-mpris
+    mlite
+  ];
+
+  postInstall = ''
+    install -Dm644 ${./org.asteroidos.btsyncd.conf} $out/share/dbus-1/system.d/org.asteroidos.btsyncd.conf
+  '';
+
+  meta = with lib; {
+    description = "Bluetooth Low Energy synchronization daemon for AsteroidOS";
+    homepage = "https://asteroidos.org/";
+    license = licenses.gpl3Plus;
+    platforms = platforms.linux;
+  };
+})
diff --git a/overlay/asteroidos/asteroid-btsyncd/org.asteroidos.btsyncd.conf b/overlay/asteroidos/asteroid-btsyncd/org.asteroidos.btsyncd.conf
new file mode 100644
index 00000000..a856e19c
--- /dev/null
+++ b/overlay/asteroidos/asteroid-btsyncd/org.asteroidos.btsyncd.conf
@@ -0,0 +1,13 @@
+<!DOCTYPE busconfig PUBLIC "-//freedesktop//DTD D-BUS Bus Configuration 1.0//EN"
+ "http://www.freedesktop.org/standards/dbus/1.0/busconfig.dtd">
+<busconfig>
+  <policy user="root">
+    <allow own="org.asteroidos.btsyncd"/>
+    <allow send_destination="org.asteroidos.btsyncd"/>
+    <allow receive_sender="org.asteroidos.btsyncd"/>
+  </policy>
+  <policy context="default">
+    <allow send_destination="org.asteroidos.btsyncd"/>
+    <allow receive_sender="org.asteroidos.btsyncd"/>
+  </policy>
+</busconfig>
diff --git a/overlay/asteroidos/asteroid-launcher/default.nix b/overlay/asteroidos/asteroid-launcher/default.nix
new file mode 100644
index 00000000..20f9f747
--- /dev/null
+++ b/overlay/asteroidos/asteroid-launcher/default.nix
@@ -0,0 +1,82 @@
+{
+  lib,
+  stdenv,
+  fetchFromGitHub,
+  cmake,
+  pkg-config,
+  qt6,
+  qml-asteroid,
+  mlite,
+  mce-dev,
+  mce-qt6,
+  ngf-qt6,
+  usb-moded-qt6,
+  keepalive,
+  timed-qt6,
+  dbus,
+  dbus-glib,
+  systemd,
+  kdePackages,
+  buildPackages,
+}:
+
+let
+  isCross = !stdenv.buildPlatform.canExecute stdenv.hostPlatform;
+in
+stdenv.mkDerivation (finalAttrs: {
+  pname = "asteroid-launcher";
+  version = "nightly";
+
+  src = fetchFromGitHub {
+    owner = "AsteroidOS";
+    repo = "asteroid-launcher";
+    rev = "4bd3d281fe3a2ea9631c640e1f00871fa19bbd4c";
+    sha256 = "194hn7pkq100339gnavvjmyib15m557bykkyqr77jrgz0wl07bpz";
+  };
+
+  strictDeps = true;
+
+  nativeBuildInputs = [
+    cmake
+    kdePackages.extra-cmake-modules
+    pkg-config
+    qt6.wrapQtAppsHook
+  ];
+
+  cmakeFlags = [
+    "-DECM_DIR=${kdePackages.extra-cmake-modules}/share/ECM/cmake"
+  ]
+  ++ lib.optionals isCross [
+    "-DQt6QmlTools_DIR=${buildPackages.qt6.qtdeclarative}/lib/cmake/Qt6QmlTools"
+    "-DQt6QuickTools_DIR=${buildPackages.qt6.qtdeclarative}/lib/cmake/Qt6QuickTools"
+    "-DQt6ShaderToolsTools_DIR=${buildPackages.qt6.qtshadertools}/lib/cmake/Qt6ShaderToolsTools"
+    "-DQt6LinguistTools_DIR=${buildPackages.qt6.qttools}/lib/cmake/Qt6LinguistTools"
+  ];
+
+  buildInputs = [
+    qt6.qtbase
+    qt6.qtdeclarative
+    qt6.qtwayland
+    qt6.qtsensors
+    qt6.qtshadertools
+    qml-asteroid
+    mlite
+    mce-dev
+    mce-qt6
+    ngf-qt6
+    usb-moded-qt6
+    keepalive
+    timed-qt6
+    dbus
+    dbus-glib
+    systemd
+  ];
+
+  meta = with lib; {
+    description = "AsteroidOS Wayland compositor and smartwatch home shell";
+    homepage = "https://asteroidos.org/";
+    license = licenses.gpl3Plus;
+    mainProgram = "asteroid-launcher";
+    platforms = platforms.linux;
+  };
+})
diff --git a/overlay/asteroidos/default.nix b/overlay/asteroidos/default.nix
index 90363944..6bddf6f8 100644
--- a/overlay/asteroidos/default.nix
+++ b/overlay/asteroidos/default.nix
@@ -7,8 +7,34 @@ rec {
 
   qml-asteroid = callPackage ./qml-asteroid { };
 
+  qt6-qpa-hwcomposer-plugin = callPackage ./qt6-qpa-hwcomposer-plugin {
+    inherit (pkgs)
+      libhybris
+      android-headers
+      glib
+      mtdev
+      ;
+    libudev = pkgs.udev;
+  };
+
+  asteroid-launcher = callPackage ./asteroid-launcher {
+    inherit
+      qml-asteroid
+      mlite
+      mce-dev
+      mce-qt6
+      ngf-qt6
+      usb-moded-qt6
+      keepalive
+      timed-qt6
+      ;
+  };
+
   amber-mpris = callPackage ./amber-mpris { };
   mlite = callPackage ./mlite { };
+  nemo-qml-plugin-configuration = callPackage ./nemo-qml-plugin-configuration {
+    inherit mlite;
+  };
   mce-dev = callPackage ./mce-dev { };
   mce-qt6 = callPackage ./mce-qt6 {
     inherit mce-dev;
@@ -18,5 +44,11 @@ rec {
   keepalive = callPackage ./keepalive { };
   timed-qt6 = callPackage ./timed-qt6 { };
 
+  asteroid-btsyncd = callPackage ./asteroid-btsyncd {
+    inherit qml-asteroid amber-mpris mlite;
+  };
+
+  virtualkeyboard = callPackage ./virtualkeyboard { };
+
   inherit (pkgs) msm-fb-refresher;
 }
diff --git a/overlay/asteroidos/nemo-qml-plugin-configuration/default.nix b/overlay/asteroidos/nemo-qml-plugin-configuration/default.nix
new file mode 100644
index 00000000..214fc9e5
--- /dev/null
+++ b/overlay/asteroidos/nemo-qml-plugin-configuration/default.nix
@@ -0,0 +1,69 @@
+{
+  lib,
+  stdenv,
+  fetchFromGitHub,
+  cmake,
+  pkg-config,
+  qt6,
+  buildPackages,
+  mlite,
+}:
+
+stdenv.mkDerivation (finalAttrs: {
+  pname = "nemo-qml-plugin-configuration";
+  version = "nightly";
+
+  src = fetchFromGitHub {
+    owner = "sailfishos";
+    repo = "nemo-qml-plugin-configuration";
+    rev = "b8ff8957f7ed9d2e3fceffc782434f2609768686";
+    hash = "sha256-KYmzafNLN+lhX45qazzZEjrEK251lOi4o6Qlas6/11g=";
+  };
+
+  strictDeps = true;
+  dontWrapQtApps = true;
+
+  nativeBuildInputs = [
+    cmake
+    pkg-config
+  ];
+
+  buildInputs = [
+    qt6.qtbase
+    qt6.qtdeclarative
+    mlite
+  ];
+
+  cmakeFlags = [
+    "-DQt6Qml_DIR=${qt6.qtdeclarative}/lib/cmake/Qt6Qml"
+    "-DQt6QmlTools_DIR=${buildPackages.qt6.qtdeclarative}/lib/cmake/Qt6QmlTools"
+  ];
+
+  postPatch = ''
+    cat > CMakeLists.txt <<'EOF'
+    cmake_minimum_required(VERSION 3.16)
+    project(nemo-qml-plugin-configuration LANGUAGES CXX)
+    set(CMAKE_CXX_STANDARD 17)
+    set(CMAKE_AUTOMOC ON)
+    find_package(Qt6 REQUIRED COMPONENTS Core Qml)
+    find_package(PkgConfig REQUIRED)
+    pkg_check_modules(MLITE REQUIRED IMPORTED_TARGET mlite6)
+    add_library(nemoconfiguration MODULE
+      src/plugin.cpp
+      src/configurationgroup.cpp
+      src/configurationvalue.cpp
+    )
+    target_include_directories(nemoconfiguration PRIVATE src ''${MLITE_INCLUDE_DIRS})
+    target_link_libraries(nemoconfiguration PRIVATE Qt6::Core Qt6::Qml PkgConfig::MLITE)
+    install(TARGETS nemoconfiguration LIBRARY DESTINATION lib/qt-6/qml/Nemo/Configuration)
+    install(FILES src/qmldir src/plugins.qmltypes DESTINATION lib/qt-6/qml/Nemo/Configuration)
+    EOF
+  '';
+
+  meta = with lib; {
+    description = "Nemo configuration QML plugin";
+    homepage = "https://github.com/sailfishos/nemo-qml-plugin-configuration";
+    license = licenses.bsd3;
+    platforms = platforms.linux;
+  };
+})
diff --git a/overlay/asteroidos/qt6-qpa-hwcomposer-plugin/default.nix b/overlay/asteroidos/qt6-qpa-hwcomposer-plugin/default.nix
new file mode 100644
index 00000000..3702fd20
--- /dev/null
+++ b/overlay/asteroidos/qt6-qpa-hwcomposer-plugin/default.nix
@@ -0,0 +1,198 @@
+{
+  lib,
+  stdenv,
+  fetchFromGitHub,
+  fetchurl,
+  runCommand,
+  cmake,
+  pkg-config,
+  qt6,
+  buildPackages,
+  libhybris,
+  android-headers,
+  libudev,
+  glib,
+  mtdev,
+}:
+
+let
+  metaAsteroid = "6c1daa5cc617cf49ad316cdc9ddc4cbd290bfbe8";
+  patch =
+    name: hash:
+    fetchurl {
+      url = "https://raw.githubusercontent.com/AsteroidOS/meta-asteroid/${metaAsteroid}/recipes-qt/qt6/qt6-qpa-hwcomposer-plugin/${name}";
+      inherit hash;
+    };
+
+  # HWC2 was introduced after the Android 6 headers in android-headers 23.
+  # Keep the old headers for HWC1/libhybris and use the Android 10 HWC2
+  # headers for this plugin.
+  hwc2Headers = runCommand "android-hwc2-headers" { } ''
+    mkdir -p $out/include/hardware
+    cp ${
+      fetchurl {
+        url = "https://raw.githubusercontent.com/aosp-mirror/platform_hardware_libhardware/android-10.0.0_r1/include/hardware/hardware.h";
+        sha256 = "1s5f4gxmqhd4xfddwaiw9q8csjnk8fcmb02370v7xj6n100jzk05";
+      }
+    } $out/include/hardware/hardware.h
+    cp ${
+      fetchurl {
+        url = "https://raw.githubusercontent.com/aosp-mirror/platform_hardware_libhardware/android-10.0.0_r1/include/hardware/hwcomposer_defs.h";
+        sha256 = "1far10f35b95apgqdasmpl7jgdqm51j5xjwzsgj3h59n5n1almhi";
+      }
+    } $out/include/hardware/hwcomposer_defs.h
+    cp ${
+      fetchurl {
+        url = "https://raw.githubusercontent.com/aosp-mirror/platform_hardware_libhardware/android-10.0.0_r1/include/hardware/hwcomposer.h";
+        sha256 = "12i9w601lnsli38wnwal81zzvfr4x6ynlxhm2crd1qwhw29jwqz7";
+      }
+    } $out/include/hardware/hwcomposer.h
+    cp ${
+      fetchurl {
+        url = "https://raw.githubusercontent.com/aosp-mirror/platform_hardware_libhardware/android-10.0.0_r1/include/hardware/hwcomposer2.h";
+        sha256 = "0f2vp1mqvn7psvx0xhdfspxjf07f4gr42wprkv1l8n5llfxqklgd";
+      }
+    } $out/include/hardware/hwcomposer2.h
+  '';
+in
+stdenv.mkDerivation (finalAttrs: {
+  pname = "qt6-qpa-hwcomposer-plugin";
+  version = "6.3.0+git";
+
+  src = fetchFromGitHub {
+    owner = "mer-hybris";
+    repo = "qt5-qpa-hwcomposer-plugin";
+    rev = "998956aebe21ac7ba6e7315d1c12e6e11c93d742";
+    hash = "sha256-AOannPAXLDEUedz6KFLvLZE92VbhgNG1vdxDRbJLhaw=";
+  };
+
+  sourceRoot = "source/hwcomposer";
+  patchFlags = [ "-p2" ];
+
+  patches = [
+    (patch "0001-Add-ambient-mode-display-support.patch" "sha256-dkORtCZzlhuIv8TxATysSsnK7fHwi+TsOdQzxwdhMoc=")
+    (patch "0003-Support-Qt-6.10.patch" "sha256-HbMYiIgzilhqQtCmr73RkX8V+aEvV8wJae6tfDB0CvQ=")
+    (patch "0005-Explicitly-make-hwcomposer-a-shared-library.patch" "sha256-Z/zjAhKm6XRQ8zVuSYPiD+rDh7KqZuKzTiUAzmsu+ZE=")
+    (patch "0006-v10-Fix-refresh-rate-fallback-unit.patch" "sha256-XOplqx7zKgbjPWIwlJ82O1qFDXXNkj2PDO/TdaP/ync=")
+    (patch "0007-v10-Make-updates-request-driven-and-vsync-paced.patch" "sha256-EQhs66R9Vc2a2yH5zVmo8xtC+oVQHjMlKyNAIGrEsbA=")
+  ];
+
+  strictDeps = true;
+  dontWrapQtApps = true;
+
+  nativeBuildInputs = [
+    cmake
+    pkg-config
+  ];
+
+  buildInputs = [
+    qt6.qtbase
+    qt6.qtdeclarative
+    qt6.qtwayland
+    libhybris
+    android-headers
+    hwc2Headers
+    libudev
+    glib
+    mtdev
+  ];
+
+  cmakeFlags = [
+    "-DQt6QmlTools_DIR=${buildPackages.qt6.qtdeclarative}/lib/cmake/Qt6QmlTools"
+    "-DQt6QuickTools_DIR=${buildPackages.qt6.qtdeclarative}/lib/cmake/Qt6QuickTools"
+  ];
+
+  postPatch = ''
+    substituteInPlace main.cpp \
+      --replace-fail 'Q_PLUGIN_METADATA(IID QPlatformIntegrationFactoryInterface_iid FILE "hwcomposer.json")' \
+                     'Q_PLUGIN_METADATA(IID "org.qt-project.Qt.QPA.QPlatformIntegrationFactoryInterface.5.3" FILE "hwcomposer.json")'
+
+    cat > CMakeLists.txt <<EOF
+    cmake_minimum_required(VERSION 3.16)
+    project(hwcomposer LANGUAGES CXX)
+    set(CMAKE_CXX_STANDARD 17)
+    set(CMAKE_AUTOMOC ON)
+
+    find_package(Qt6 REQUIRED COMPONENTS Core Gui DBus OpenGL WaylandCompositor)
+    find_package(PkgConfig REQUIRED)
+    pkg_check_modules(GLIB REQUIRED glib-2.0)
+    pkg_check_modules(LIBUDEV REQUIRED libudev)
+    pkg_check_modules(MTDEV REQUIRED mtdev)
+
+    add_library(hwcomposer MODULE
+      main.cpp
+      hwcomposer_context.cpp
+      hwcomposer_screeninfo.cpp
+      hwcomposer_backend.cpp
+      hwcomposer_backend_v0.cpp
+      hwcomposer_backend_v10.cpp
+      hwcomposer_backend_v11.cpp
+      hwcomposer_backend_v20.cpp
+      qeglfsbackingstore.cpp
+      qeglfscontext.cpp
+      qeglfsintegration.cpp
+      qeglfsscreen.cpp
+      qeglfswindow.cpp
+    )
+
+    target_include_directories(hwcomposer PRIVATE
+      .
+      ${hwc2Headers}/include
+      ${android-headers}/include/android
+      ${android-headers}/include/android/hardware
+      ${libhybris}/include
+      ${libhybris}/include/hybris/eglplatformcommon
+      ${libhybris}/include/hybris/hwcomposerwindow
+      ${qt6.qtbase}/include/QtCore/${qt6.qtbase.version}
+      ${qt6.qtbase}/include/QtCore/${qt6.qtbase.version}/QtCore
+      ${qt6.qtbase}/include/QtGui/${qt6.qtbase.version}
+      ${qt6.qtbase}/include/QtGui/${qt6.qtbase.version}/QtGui
+      ${qt6.qtbase}/include/QtGui/${qt6.qtbase.version}/QtGui/private
+      ${qt6.qtbase.dev}/include/QtCore/${qt6.qtbase.version}
+      ${qt6.qtbase.dev}/include/QtGui/${qt6.qtbase.version}
+      ${qt6.qtbase.dev}/include/QtGui/${qt6.qtbase.version}/QtGui
+      ${qt6.qtwayland.dev}/include/QtWaylandCompositor/${qt6.qtwayland.version}
+      ${qt6.qtwayland.dev}/include/QtWaylandCompositor/${qt6.qtwayland.version}/QtWaylandCompositor
+      ''${GLIB_INCLUDE_DIRS}
+      ''${LIBUDEV_INCLUDE_DIRS}
+      ''${MTDEV_INCLUDE_DIRS}
+    )
+    target_compile_definitions(hwcomposer PRIVATE
+      QEGL_EXTRA_DEBUG
+      MESA_EGL_NO_X11_HEADERS
+      QT_OPENGL_ES_2
+      GL_GLEXT_PROTOTYPES
+      HWC_PLUGIN_HAVE_HWCOMPOSER1_API
+      HWC_PLUGIN_HAVE_HWCOMPOSER2_API
+    )
+    target_compile_options(hwcomposer PRIVATE
+      "SHELL:-include stddef.h"
+      "SHELL:-include ${hwc2Headers}/include/hardware/hardware.h"
+      "SHELL:-include ${hwc2Headers}/include/hardware/hwcomposer_defs.h"
+      "SHELL:-include ${hwc2Headers}/include/hardware/hwcomposer.h"
+    )
+    target_link_libraries(hwcomposer PRIVATE
+      Qt6::Core
+      Qt6::Gui
+      Qt6::DBus
+      Qt6::OpenGL
+      Qt6::WaylandCompositor
+      ''${GLIB_LIBRARIES}
+      ''${LIBUDEV_LIBRARIES}
+      ''${MTDEV_LIBRARIES}
+      ${libhybris}/lib/libhardware.so.2.0.0
+      ${libhybris}/lib/libhwc2.so.1.0.0
+      ${libhybris}/lib/libhybris-eglplatformcommon.so.1.0.0
+    )
+
+    install(TARGETS hwcomposer LIBRARY DESTINATION lib/qt-6/plugins/platforms)
+    EOF
+  '';
+
+  meta = with lib; {
+    description = "Qt6 QPA plugin for libhybris hwcomposer";
+    homepage = "https://github.com/mer-hybris/qt5-qpa-hwcomposer-plugin";
+    license = licenses.lgpl21Only;
+    platforms = platforms.linux;
+  };
+})
diff --git a/overlay/asteroidos/virtualkeyboard/default.nix b/overlay/asteroidos/virtualkeyboard/default.nix
new file mode 100644
index 00000000..f5df8828
--- /dev/null
+++ b/overlay/asteroidos/virtualkeyboard/default.nix
@@ -0,0 +1,52 @@
+{
+  lib,
+  stdenv,
+  fetchFromGitHub,
+  cmake,
+  pkg-config,
+  qt6,
+  buildPackages,
+}:
+
+stdenv.mkDerivation (finalAttrs: {
+  pname = "asteroid-virtualkeyboard";
+  version = "2.1.0";
+
+  src = fetchFromGitHub {
+    owner = "AsteroidOS";
+    repo = "asteroid-virtualkeyboard";
+    rev = "666743fc42eaa02b9f82155ff3aff3a40e0d45ac";
+    sha256 = "0yw1xq315p7q7illjpq92cs0060lrw94bfk8ywc7fbrr1gmkmvjf";
+  };
+
+  strictDeps = true;
+
+  nativeBuildInputs = [
+    cmake
+    pkg-config
+    qt6.wrapQtAppsHook
+  ];
+
+  buildInputs = [
+    qt6.qtbase
+    qt6.qtdeclarative
+    qt6.qtvirtualkeyboard
+  ];
+
+  cmakeFlags = [
+    "-DCMAKE_INSTALL_LIBDIR=lib"
+    "-DCMAKE_INSTALL_DATADIR=share"
+  ]
+  ++ lib.optionals (!stdenv.buildPlatform.canExecute stdenv.hostPlatform) [
+    "-DQt6QuickTools_DIR=${buildPackages.qt6.qtdeclarative}/lib/cmake/Qt6QuickTools"
+    "-DQt6QmlTools_DIR=${buildPackages.qt6.qtdeclarative}/lib/cmake/Qt6QmlTools"
+    "-DQt6ShaderToolsTools_DIR=${buildPackages.qt6.qtshadertools}/lib/cmake/Qt6ShaderToolsTools"
+  ];
+
+  meta = with lib; {
+    description = "AsteroidOS on-screen virtual keyboard with watch style and layouts";
+    homepage = "https://asteroidos.org/";
+    license = licenses.gpl3Plus;
+    platforms = platforms.linux;
+  };
+})
